### ISSUE TITLE
Cannot update PowerShell through winget because newer versions aren't always available through it

### ISSUE THREAD

**Issue by Haraven** [PowerShell/PowerShell#19075]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

The problem I'm having is that even though PowerShell notifies me that there's a new version available (in this case, 7.3.2), I can't upgrade through the `winget` application, because winget only has my current version (7.3.1) available. As a consequence, running `winget upgrade Microsoft.PowerShell` tells me the application is already up to date.

### Expected behavior

New versions of PowerShell should always and immediately be available through winget as well, if an update notification goes out. 

I say _should_ because winget is [listed as the recommended method for installing PowerShell on Windows](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows), but if it doesn't have the latest version of PowerShell available, it defeats the purpose of making that recommendation. 

What's more, even the official guide for installing PowerShell implies, at least at the time of writing, that 7.3.1 is the latest stable version of PowerShell, even though it should say it's 7.3.2.

### Actual behavior

`winget upgrade Microsoft.PowerShell` says there is no applicable upgrade, even though a new version of PowerShell is available.

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.1
PSEdition                      Core
GitCommitId                    7.3.1
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

Here is a screenshot of my PowerShell terminal tab that clearly shows a) a new version notification, b) winget not having that new version, c) winget upgrade not installing the new PowerShell version:

![Screenshot_20230201_091153](https://user-images.githubusercontent.com/10146027/215977549-229f59b4-56ab-4b4d-8d46-fc34db36af3d.png)

**Comment by DEAN-Cherry** [comment]
same for me

**Comment by J0F3** [comment]
same. 

The links in the documentation to the MSI packages is not updated also. (still version 7.3.1)

**Comment by mgreenegit** [comment]
The PR is open here. There was unfortunately a delay in submitting it.
https://github.com/microsoft/winget-pkgs/pull/95159

**Comment by atticusrussell** [comment]
same here. even though on the MS website it lists 7.3.2 in an example
https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.3

**Comment by jhoneill** [comment]
> New versions of PowerShell should always and immediately be available through winget as well, if an update notification goes out.

How about we turn that on it's head ? 
The update notification should not go out until winget, store and other distribution mechanisms have are ready to deliver it. 

I'm not sure how the check is done today, but it can't be impossible to check something which is updated only when the package is downloadable from all sources.

**Comment by fluffynuts** [comment]
I've literally just updated powershell _via vscode_ because it offered to when I opened it and I've been unable to winget the update it's nagging about for days.

So another way we could turn this on it's head is if pwsh could update itself - I mean, if vscode can download & run the installer, pwsh could.

**Comment by daxian-dbw** [comment]
The PR https://github.com/microsoft/winget-pkgs/pull/95159 was merged. It was delayed due to a pipeline issue in the WinGet repository. See the comment https://github.com/microsoft/winget-pkgs/pull/95159#issuecomment-1416383365. I will close this issue as resolved.

**Comment by fluffynuts** [comment]
Thanks @daxian-dbw

**Comment by geronimus** [comment]
Although closed as resolved, 3 weeks ago at time of writing, this issue still appears to be a problem.
![2023-03-04T120313_pwsh732](https://user-images.githubusercontent.com/12959360/222896461-a14c8089-f909-4ac3-ad3e-9b484ad6aa52.png)

**Comment by jhoneill** [comment]
It's showing up for me. 

<img width="651" alt="image" src="https://user-images.githubusercontent.com/16209593/222918856-7590584d-6f89-4507-984b-2b03f96a8a2d.png">

**Comment by geronimus** [comment]
@jhoneill You're quite right!
But why then is it not showing up as an upgradable package with the command `winget upgrade`?

**Comment by geronimus** [comment]
Never mind. I think I've found the answer:
![2023-03-04T183112_pwsh732](https://user-images.githubusercontent.com/12959360/222920358-c869d766-05d9-4779-8107-ddd0fe4764c5.png)

**Comment by pcbeard** [comment]
I'm seeing this again with current installed version 7.3.5, and announced newer version 7.3.6.

**Comment by prlcutting** [comment]
Same with `7.3.7` to `7.3.8`. It seemed to take a few days for `winget` to catch up with the latest version of PowerShell after the banner in the terminal showed up saying `7.3.8` was available.

**Comment by marcin93** [comment]
Hey @prlcutting, try restarting powershell and `winget upgrade Microsoft.PowerShell` again. I was working on 7.3.6 and it took me to 7.3.7 then 7.3.8, but I did upgrade twice. From beginning I've seen 7.3.8 under `winget search Microsoft.PowerShell` as current version.

**Comment by fluffynuts** [comment]
Today I've seen powershell being updated as part of Windows Update - which is great! So how can we turn off the annoying message then, since WU will get to the upgrade eventually?

**Comment by bryanhunwardsen** [comment]
Powershell and Visual Studio Code should not notify of a powershell upgrade when it is not yet available with winget, especially when the installed powershell is managed strictly by winget. This needs to be reopened and fixed permanently as this behaior has been repreated multiple times this calendar year.

![image](https://github.com/PowerShell/PowerShell/assets/3408302/0c2907e0-79be-4e10-bb53-a016436ae71c)

**Comment by smchargue** [comment]
This is just annoying and seems so unnecessary, why are we getting nagged to update something that isn't available yet?  I agree that if winget is the *recommended* approach, then it **should** behave properly as the OP stated. 

And why is this closed when it clearly is not resolved. 

![image](https://github.com/PowerShell/PowerShell/assets/4064621/ea5f3772-5cc0-4d7b-bd62-094f9ef31ca9)

**Comment by kilasuit** [comment]
Addressing some of the comments in this issue

> It seemed to take a few days for winget to catch up with the latest version of PowerShell after the banner in the terminal showed up saying 7.3.8 was available.

Unfortunately this is something that will be for a while as this is not fully automated to publish to [winget-pkgs repo](https://github.com/microsoft/winget-pkgs/) yet.

Even when that is done there are then barriers in WinGet's approval process to their feed that will stop it being listed right away and likely will never be available instantly **unless** they add a [mechanism for you to search for & install packages that are pending to be added to the feed](https://github.com/microsoft/winget-cli/issues/3443#issuecomment-1640537859), but that's a winget issue not a PowerShell issue to fix

> Powershell and Visual Studio Code should not notify of a powershell upgrade when it is not yet available with winget, especially when the installed powershell is managed strictly by winget. 

> Today I've seen powershell being updated as part of Windows Update - which is great! So how can we turn off the annoying message then, since WU will get to the upgrade eventually?

Just because a package was installed with one method doesn't mean it is then managed by that method so it can't be assumed to be the case however the notifications (in the console and in VSCode) can be disabled by configuration options already by users or admins easily enough and are documented already.
PowerShell can be done by setting the `POWERSHELL_UPDATECHECK` environment vairable to `Off` as per the [about_update_notifications help doc](https://learn.microsoft.com/en-gb/powershell/module/microsoft.powershell.core/about/about_update_notifications?view=powershell-7.3) 
VSCode can be done with this setting `"powershell.promptToUpdatePowerShell": false` 

There is the possible ask here - `that when installed from xyz method, run an post-install task to set those settings if possible` however my view (not the teams btw) is that this should be managed either by the end user/admin/org administrator as otherwise if each install method sets it then you could have the expectation to see the message and set it so you do and then the process overwrites your settings.

> This is just annoying and seems so unnecessary, why are we getting nagged to update something that isn't available yet? 

Well understood to the team & community how annoying it can be with how this currently works but it is of use for those of us that aren't using any package management tooling & area installing as soon as they can directly from GitHub when it is released or those that have many different ways of managing their software on their machines all working together (or like I found at one point [perhaps against each other like I found in [PowerShell Update Oddness](https://blog.kilasuit.org/2023/03/23/powershell-update-oddness/))

However technically it is available, just not **immediately** available via certain install/update methods like winget, chocolatey, Microsoft Update etc which all have delays due to having their own schedules for verifying and then publishing the package to their feeds & is mostly outside the PSTeam responsibility to improve.

However, the team is working on getting this process more streamlined so that when the latest pwsh version is available [the work to push to winget's package repo will be automated so that it can be made available in winget as soon as they make it possible](https://github.com/PowerShell/PowerShell/pull/20563) as it's still done manually today. As mentioned there are also however additional steps required by winget team that slows down packages from being published to their feed as can be seen from the Azure Pipelines being run in the [PR to add v.7.3.9 to winget's feed](https://github.com/microsoft/winget-pkgs/pull/123901) 

Again this is a winget area to solve more than a PowerShell Team one to solve but they are working together to shorten the gap as best as possible.

However if winget did add ability for searching for packages that are pending approval & listing in the feed then that would negate for lots of similar issues and comments around this being raised here and other software repos going forward. Hence why I'd [raised this issue to ask for the winget-cli team to add a way to do it](https://github.com/microsoft/winget-cli/issues/3443) - please upvote it if you think it makes sense.

> So another way we could turn this on it's head is if pwsh could update itself - I mean, if vscode can download & run the installer, pwsh could.

This has been asked for time and time again and whilst it's been asked for a built in Update-PowerShell cmdlet the closest that I have found is via the Install-PowerShell script available in this repo which I use as a way for [installing side by side versions for testing purposes](https://blog.kilasuit.org/2023/09/09/installing-powershell-side-by-side-for-testing-purposes/)
Whilst it isn't currently a documented way for installing PowerShell it is in use in many places and does still give you a supported install of PowerShell at the end of it.

Hope all these comments helped understand what the latest issue is.

**Comment by fluffynuts** [comment]
Thanks - the real win for me is learning about `POWERSHELL_UPDATECHECK` and simply setting that to `Off`, system-wide. I don't need to have the latest version as it drops and I'm quite happy to let, eg, windows update (or my sparodic `winget upgrade --all` invocations) do the work. I just wanted to suppress the message.

**Comment by fogolin** [comment]
Ok, so I was facing the same issue as you guys, I had version 7.3.9 installed, and v 7.3.10 was available:

![image](https://github.com/PowerShell/PowerShell/assets/11749606/76fd5da6-37ec-4c8d-b6e6-fb2f1eb5ee22)

After, forcing the installation, it seems the package was downloaded from the correct version build:
![image](https://github.com/PowerShell/PowerShell/assets/11749606/975b933d-a665-4c66-a65b-945aa3ee7b80)

Even thought it says there is an upgrade available, now my Powershell version is the correct latest one:
![image](https://github.com/PowerShell/PowerShell/assets/11749606/628d9a82-e5b5-438d-89d6-4ba01259f6fe)

Seems like an indexing issue rather than a problem with winget, but forcing the installation was enough to set the latest version of the software on my machine.

**Comment by Peyman-98** [comment]
![image](https://github.com/PowerShell/PowerShell/assets/12208479/31b21c59-6de2-49a3-a19b-75b79785a62b)
what should i do abouth this?

**Comment by fogolin** [comment]
Same here:
![image](https://github.com/PowerShell/PowerShell/assets/11749606/f1846b2a-10c9-4fd3-81a0-ad691c0f0a52)

**Comment by metamec** [comment]
We are back to square one with the release of v7.4.3.  

I just set the environment variable POWERSHELL_UPDATECHECK=Off to suppress the update nagging.

I can't be bothered with this nonsense every few months.

You gave us winget, Microsoft.  Start supporting it fully, please.

**Comment by jimmy-zhening-luo** [comment]
I am still having this issue as of v7.4.3. I receive the nag prompt:

```
PowerShell 7.4.2

   A new PowerShell stable release is available: v7.4.3
   Upgrade now, or check out the release page at:
     https://aka.ms/PowerShell-Release?tag=v7.4.3
```

Yet `winget` gives me no recourse:

```bash
winget search Microsoft.Powershell
Name               Id                           Version Source
---------------------------------------------------------------
PowerShell         Microsoft.PowerShell         7.4.2.0 winget
PowerShell Preview Microsoft.PowerShell.Preview 7.5.0.3 winget
```

The version can't be found through invoking 'install' either, whether or not I specify an explicit version of 7.4.3:

```bash
winget install --id Microsoft.Powershell --source winget
Found an existing package already installed. Trying to upgrade the installed package...
No version found matching: 7.4.3
```

```bash
winget install --id Microsoft.Powershell -v 7.4.3 --source winget
Found an existing package already installed. Trying to upgrade the installed package...
No version found matching: 7.4.3
```

If I were to scour the entire canonical [installation MSDN doc](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-7.4), it is clear that the well-supported path is intended to be `winget`.

Were I to try to remediate by finding an `.msi`, I would have to have clicked the [nag-link](https://github.com/PowerShell/PowerShell/releases/tag/v7.4.3) and happen to know by convention that I can find various .msi flavors at the bottom of the release. Thankfully I did!

As a proud former Windows employee I'm a bit embarrassed. Pls fix.
